pkg/cloudprovider/vsphere/nodemanager.go

```diff
@@ -386,6 +333,169 @@ func (nm *NodeManager) DiscoverNode(nodeID string, searchBy cm.FindVM) error {
 	return nil
 }
 
+func discoverIPs(ipAddrNetworkNames []*ipAddrNetworkName, ipFamily string,
+	internalNetworkSubnet, externalNetworkSubnet *net.IPNet,
+	internalVMNetworkName, externalVMNetworkName string) (internal *ipAddrNetworkName, external *ipAddrNetworkName) {
+
+	ipFamilyMatches := findMatchesForIPFamily(ipAddrNetworkNames, ipFamily)
+
+	var discoveredInternal *ipAddrNetworkName
+	var discoveredExternal *ipAddrNetworkName
+
+	if len(ipFamilyMatches) != 0 {
+		discoveredInternal = findSubnetMatch(ipFamilyMatches, internalNetworkSubnet)
+		if discoveredInternal != nil {
+			klog.V(2).Infof("Adding Internal IP by AddressMatching: %s", discoveredInternal.ipAddr)
+		}
+		discoveredExternal = findSubnetMatch(ipFamilyMatches, externalNetworkSubnet)
+		if discoveredExternal != nil {
+			klog.V(2).Infof("Adding External IP by AddressMatching: %s", discoveredExternal.ipAddr)
+		}
+
+		if discoveredInternal == nil && internalVMNetworkName != "" {
+			discoveredInternal = findNetworkNameMatch(ipFamilyMatches, internalVMNetworkName)
+			if discoveredInternal != nil {
+				klog.V(2).Infof("Adding Internal IP by NetworkName: %s", discoveredInternal.ipAddr)
+			}
+		}
+
+		if discoveredExternal == nil && externalVMNetworkName != "" {
+			discoveredExternal = findNetworkNameMatch(ipFamilyMatches, externalVMNetworkName)
+			if discoveredExternal != nil {
+				klog.V(2).Infof("Adding External IP by NetworkName: %s", discoveredExternal.ipAddr)
+			}
+		}
+
+		// At least one of the Internal or External addresses has been found.
+		// Minimally the Internal needs to exist for the node to function correctly.
+		// If only one was discovered, will log the warning and continue which will
+		// ultimately be visible to the end user
+		if discoveredInternal != nil || discoveredExternal != nil {
+			if discoveredInternal != nil && discoveredExternal == nil {
+				klog.Warning("Internal address found, but external address not found. Returning what addresses were discovered.")
+			} else if discoveredInternal == nil && discoveredExternal != nil {
+				klog.Warning("External address found, but internal address not found. Returning what addresses were discovered.")
+			}
+		}
+
+		// Neither internal or external addresses were found. This defaults to the legacy
+		// address selection behavior which is to only support a single address and
+		// return the first one found
+		if discoveredInternal == nil && discoveredExternal == nil {
+			klog.V(5).Info("Default address selection. Single NIC, Single IP Address")
+			klog.V(2).Infof("Adding IP: %s", ipFamilyMatches[0].ipAddr)
+			discoveredInternal = ipFamilyMatches[0]
+			discoveredExternal = ipFamilyMatches[0]
+		}
```

Comment:
this can be refactored a little I think:
```
if discoveredInternal == nil && discoveredExternal == nil {
   // Line 385-388
} else {
   // Line 374-378
}
```

Reply:
Good eye, we've updated the PR to your suggestion.